File: 0494-target-sum/0494-target-sum.java
class Solution {
    public int permute(int idx, int[] nums, int target){
        if(idx == nums.length && (target == 0)){
            return 1;
        }
        if(idx>=nums.length) return 0;

        int add = permute(idx+1, nums, target+nums[idx]);
        int sub = permute(idx+1, nums, target-nums[idx]);

        return add+sub;
    }

    public int findTargetSumWays(int[] nums, int target) {
        return permute(0, nums, target);
    }
}

File: 0494-target-sum/0494-target-sum.py
class Solution(object):
    def findTargetSumWays(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        n = len(nums)
        def helper(idx, sumSoFar, dp):
            if idx==n:
                if sumSoFar == target:
                    return 1
                return 0

            if (idx, sumSoFar) in dp:
                return dp[(idx, sumSoFar)]

            #Add
            add = helper(idx+1, sumSoFar+nums[idx], dp)

            #Subtract
            sub = helper(idx+1, sumSoFar-nums[idx], dp)
            
            dp[(idx, sumSoFar)] = add + sub
            return add + sub 

        dp = {}
        return helper(0, 0, dp)